### PR TITLE
Upload large dataframes in chunks

### PR DESCRIPTION
**Pull request by ian-contiamo**
@nagautier @bmwilly can you please double-check the formula I used to slice the data? We need to be absolutely certain that we do not skip (or repeat) a single row!!! I'm 99.9% certain of what I did, even in edge cases, but I'd like to add a few more 9s ;)

### CODE REVIEW

**Comment by ian-contiamo** [comment]
@bmwilly The test that fails is unrelated to this commit. It does not fail for me locally. Not sure how to address this.

**Comment by bmwilly** [comment]
For posterity, the tests were failing because they're hitting a live server. See discussion issue: https://github.com/contiamo/contiamo-client-python/issues/26

File: contiamo/data.py [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
+
 import numpy
 import pandas as pd
-import tempfile
 
 from contiamo.http_client import HTTPClient
 from contiamo.utils import contract_url_template_from_identifier, raise_response_error
 from contiamo.utils import get_file_extension
-from contiamo.errors import InvalidRequestError
+from contiamo.errors import ContiamoException, InvalidRequestError
```
Comment (bmwilly) [inline contiamo/data.py]: I know it's not part of this PR, but let's use this opportunity to alphabetize this import section, and we only need to import `from contiamo.utils` once.

File: contiamo/data.py [hunk 2]
```diff
@@ -137,10 +139,31 @@ def discover(self, dataframe=None, filename=None, include_index=False):
...
+                try:
+                    # we ignore the response, all that matters is the response code
+                    # anything other than a 200 will raise an error
+                    self._post_data(url, tmp, filename, include_index)
+                except ContiamoException as e:
+                    warnings.warn(
+                        'Request #%d has failed, %d rows have been uploaded so far.'
+                        % (n+1, n*chunk_size)
+                    )
+                    raise
```
Comment (bmwilly) [inline contiamo/data.py]: `raise e` is more informative

File: contiamo/data.py [hunk 3]
```diff
@@ -137,10 +139,31 @@ def discover(self, dataframe=None, filename=None, include_index=False):
         url = self.url_template.format(action='upload/discover')
         return self._post_data(url, dataframe, filename, include_index)
 
-    def upload(self, dataframe=None, filename=None, include_index=False):
+    def upload(self, dataframe=None, filename=None, include_index=False, chunk_size=100000):
```
Comment (bmwilly) [inline contiamo/data.py]: as discussed, let's default to `len(dataframe)`

File: contiamo/tests/test_requests.py [hunk 4]
```diff
@@ -107,3 +107,8 @@ def test_two_arguments(self):
         df = pd.DataFrame({'a': [1, 2, 3], 'b': [4, 5, 6]})
         with pytest.raises(InvalidRequestError):
             data_client.upload(dataframe=df, filename=utils.file_test_data('mock_data.csv'))
+
+    def test_invalid_chunk_size(self):
```
Comment (bmwilly) [inline contiamo/tests/test_requests.py]: this test is unnecessary IMO. A test for the chunking / data slicing is probably the only new functionality that needs to be tested (not saying you have to, just if you wanted to add more tests).

Comment (ian-contiamo) [reply contiamo/tests/test_requests.py]: It's actually a regression test in disguise: I just want to make sure no-one removes the functionality and hence the `chunk_size` parameter by mistake (in which case a `TypeError` would be raised).

Comment (ian-contiamo) [reply contiamo/tests/test_requests.py]: I didn't add a proper test because it was a bit impractical, but I've thought of a couple of ways to do so over lunch...

File: contiamo/data.py (hunk 2 above)
Comment (ian-contiamo) [reply contiamo/data.py]: If you say so ;)

File: contiamo/data.py (hunk 3 above)
Comment (ian-contiamo) [reply contiamo/data.py]: Is `None` or `0` a better default argument? (I'm hijacking this PR to discuss fine Python nuances... ;)

File: contiamo/data.py (hunk 1 above)
Comment (ian-contiamo) [reply contiamo/data.py]: Good catch (the duplicate `utils` import).

Notice how I separated into standard imports, external imports, and internal ones already (with two out of three alphabetized!)

Comment (bmwilly) [reply contiamo/data.py]: yes, thank you!

File: contiamo/data.py (hunk 3 above)
Comment (bmwilly) [reply contiamo/data.py]: I think `None`. That often implies that some default will be used.

File: contiamo/tests/test_requests.py (hunk 4 above)
Comment (bmwilly) [reply contiamo/tests/test_requests.py]: >It's actually a regression test in disguise: I just want to make sure no-one removes the functionality and hence the chunk_size parameter by mistake (in which case a TypeError would be raised).

fair enough 

>I didn't add a proper test because it was a bit impractical, but I've thought of a couple of ways to do so over lunch... What do you think?

yeah, I thought it might be more trouble than it's worth, so happy to leave this decision to you :)

**Comment by bmwilly** [comment]
lgtm pending the small imports improvement ;)

**Comment by ian-contiamo** [comment]
@bmwilly can you check this commit? 
https://github.com/contiamo/contiamo-client-python/pull/24/commits/5b8c0731d8e2f3aaf856916000e784252d03acad

File: contiamo/data.py [hunk 5]
```diff
@@ -52,6 +54,13 @@ def preformat(df):
         df[col] = df.loc[df[col].notnull(), col].astype(int).astype(str)
 
 
+def slice_in_chunks(length, chunk_size):
```
Comment (bmwilly) [inline contiamo/data.py]: 👍

File: contiamo/tests/test_requests.py [hunk 6]
```diff
@@ -91,6 +92,13 @@ def test_query(self):
         response = query(query_id, api_base=api_base)
         assert isinstance(response, pd.DataFrame)
 
+    def test_slice_in_chunks(self):
```
Comment (bmwilly) [inline contiamo/tests/test_requests.py]: 👍

**Comment by bmwilly** [comment]
>It makes chunking testable. And the code a bit nicer (the slicing is abstracted from the main function) 
good stuff!
